src/cudamatrix/cu-kernels.cu

```diff
@@ -2630,6 +2630,204 @@ static void _diff_log_softmax(const MatrixDim in_deriv_dim,
 }
 
 
+/// This CUDA kernel implements the idea 'spatial smoothing' as described in
+/// the paper https://arxiv.org/pdf/1610.05256v1.pdf
+///
+/// It adds spatial regularization derivative to `out_deriv`.
+/// The regularization term is calculated from `out_value`.
+/// Each row of `out_value` is first viewed as a n*16 2D image,
+/// high-pass filtered by a specific 3x3 Filter kernel with 1 at the center
+/// and -1/8 elsewhere for twice, scaled with `-scale` and then added to
+/// `out_deriv`.
+/// out_deriv = out_deriv - scale * Filter (X) (Filter (X) out_value),
```

Comment:
@danpovey Sorry for the delay. Filtering twice makes things more complicated. I just finish both the GPU and CPU implementation draft. I'm working on the testing code. Could you help check the backprop formula at this line. I'm currently not sure if it correct. Thanks!